File: 2024/3/3.pl
#!/usr/bin/env swipl
% usage: swipl 3.pl [FILENAME]

:- module(aoc2024day3, []).
:- module(aoc2024day3).
:- initialization(main, main).
:- use_module(library(clpfd)).
:- use_module(library(dcg/basics)).
:- use_module(library(dcg/high_order)).

integer999(X) --> integer(X), { X in 0..999 }.

load_token(Token) --> string(_), token(Token).

token(mul(A, B)) --> "mul(", integer999(A), ",", integer999(B), ")".
token(do)        --> "do()".
token(dont)      --> "don't()".

load(Input) --> sequence(load_token, Input), remainder(_).

part_one --> include([X] >> functor(X, mul, 2)), part_two.

part_two(Input, Result) :- phrase(interpret(1, 0, Result), Input).

interpret(P, R0, R) --> [mul(A, B)], { R1 #= R0 + A * B * P }, interpret(P, R1, R).
interpret(_, R0, R) --> [do],                                  interpret(1, R0, R).
interpret(_, R0, R) --> [dont],                                interpret(0, R0, R).
interpret(_, R,  R) --> eos.

main([File]) :-
    once(phrase_from_file(load(Input), File)),
    part_one(Input, ResultOne),
    writeln(ResultOne),
    part_two(Input, ResultTwo),
    writeln(ResultTwo).

main([]) :- main(['input.txt']).

File: 2024/3/3.jl
#!/usr/bin/env julia
# usage: julia 3.jl [FILENAME]

module AOC2024Day3

partone(input) =
    sum(prod(parse.(Int, m)) for m in eachmatch(r"mul\((\d{1,3}),(\d{1,3})\)", input))

parttwo(input) =
    partone(replace(input, r"don't\(\)(.|\s)*?(do\(\)|$)"=>""))

if abspath(PROGRAM_FILE) == @__FILE__
    input = readchomp(get(ARGS, 1, joinpath(@__DIR__, "input.txt")))
    println(partone(input))
    println(parttwo(input))
end

end # module
